Interaction regions don't update when event listeners change without rendering updates

hortont424 opened the pull request:
#### 82f024ac0b5757189369a944edc20f08c4b7a1ef
<pre>
Interaction regions don&apos;t update when event listeners change without rendering updates
<a href="https://bugs.webkit.org/show_bug.cgi?id=248804">https://bugs.webkit.org/show_bug.cgi?id=248804</a>

Reviewed by Aditya Keerthi.

* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt: Added.
* LayoutTests/interaction-region/click-handler-dynamically-added.html: Added.
* LayoutTests/interaction-region/click-handler-dynamically-removed-expected.txt: Added.
* LayoutTests/interaction-region/click-handler-dynamically-removed.html: Added.
Add some tests for both dynamic addition and removal of event listeners.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::httpEquivPolicy const):
Move this so it&apos;s not in the middle of event-related code.

(WebCore::Document::didAddOrRemoveMouseEventHandler):
Add a method that invalidates style when mouse event handlers change.

* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventNames.h:
(WebCore::EventNames::isMouseClickRelatedEventType const):
(WebCore::EventNames::isMouseMoveRelatedEventType const):
Add methods for click and move related events.

* Source/WebCore/dom/Node.cpp:
(WebCore::tryAddEventListener):
(WebCore::tryRemoveEventListener):
(WebCore::Node::willRespondToMouseMoveEvents const):
(WebCore::Node::willRespondToMouseClickEventsWithEditability const):
(WebCore::Node::willRespondToMouseWheelEvents const): Deleted.
Adopt EventNames methods instead of hardcoded name checks.
Call didAddOrRemoveMouseEventHandler when relevant.

* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::willRespondToMouseWheelEvents const): Deleted.
* Source/WebCore/html/HTMLElement.h:
Remove the unused willRespondToMouseWheelEvents.

Canonical link: <a href="https://commits.webkit.org/257453@main">https://commits.webkit.org/257453@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1374155e1cb8b5a25fe0aab077a22e187ec385b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8209 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32158 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108410 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168662 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102965 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8762 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85549 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91527 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106369 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104723 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90217 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33659 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21562 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2111 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2015 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45469 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5129 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6978 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42553 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3429 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7274,6 +7274,23 @@ unsigned Document::touchEventHandlerCount() const
+void Document::didAddOrRemoveMouseEventHandler(Node& node)
+{
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+    if (is<Element>(node)) {
```
`dynamicDowncast` 😄

hortont424 replied:
I will fix the copied-from code too :)

pxlcoder commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7136,9 +7145,9 @@ void Document::wheelEventHandlersChanged(Node* node)
 
 #if ENABLE(WHEEL_EVENT_REGIONS)
-    if (is<Element>(node)) {
+    if (Element* element = dynamicDowncast<Element>(node)) {
```
Could use `auto*`.

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7274,6 +7274,23 @@ unsigned Document::touchEventHandlerCount() const
+void Document::didAddOrRemoveMouseEventHandler(Node& node)
+{
+#if ENABLE(INTERACTION_REGIONS_IN_EVENT_REGION)
+    if (Element* element = dynamicDowncast<Element>(node)) {
```
auto*

cdumez commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2506,7 +2510,9 @@ bool Node::willRespondToMouseMoveEvents() const
 #endif
     auto& eventNames = WebCore::eventNames();
-    return hasEventListeners(eventNames.mousemoveEvent) || hasEventListeners(eventNames.mouseoverEvent) || hasEventListeners(eventNames.mouseoutEvent);
+    return eventTypes().containsIf([&](const auto& type) {
```
auto should suffice. If I understand correctly, this resolves to an enum type so no need to pass by const ref.

hortont424 replied:
It's an AtomString

cdumez commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2545,16 +2551,11 @@ bool Node::willRespondToMouseClickEventsWithEditability(Editability editability)
-{
-    return hasEventListeners(eventNames().mousewheelEvent);
+    auto& eventNames = WebCore::eventNames();
+    return eventTypes().containsIf([&](const auto& type) {
```
ditto: `auto`

hortont424 committed:
Interaction regions don't update when event listeners change without rendering updates
https://bugs.webkit.org/show_bug.cgi?id=248804

Reviewed by Aditya Keerthi.

* LayoutTests/interaction-region/click-handler-dynamically-added-expected.txt: Added.
* LayoutTests/interaction-region/click-handler-dynamically-added.html: Added.
* LayoutTests/interaction-region/click-handler-dynamically-removed-expected.txt: Added.
* LayoutTests/interaction-region/click-handler-dynamically-removed.html: Added.
Add some tests for both dynamic addition and removal of event listeners.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::httpEquivPolicy const):
Move this so it's not in the middle of event-related code.

(WebCore::Document::didAddOrRemoveMouseEventHandler):
Add a method that invalidates style when mouse event handlers change.

* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/EventNames.h:
(WebCore::EventNames::isMouseClickRelatedEventType const):
(WebCore::EventNames::isMouseMoveRelatedEventType const):
Add methods for click and move related events.

* Source/WebCore/dom/Node.cpp:
(WebCore::tryAddEventListener):
(WebCore::tryRemoveEventListener):
(WebCore::Node::willRespondToMouseMoveEvents const):
(WebCore::Node::willRespondToMouseClickEventsWithEditability const):
(WebCore::Node::willRespondToMouseWheelEvents const): Deleted.
Adopt EventNames methods instead of hardcoded name checks.
Call didAddOrRemoveMouseEventHandler when relevant.

* Source/WebCore/dom/Node.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::willRespondToMouseWheelEvents const): Deleted.
* Source/WebCore/html/HTMLElement.h:
Remove the unused willRespondToMouseWheelEvents.

Canonical link: https://commits.webkit.org/257453@main
